### ISSUE TITLE
CompletableFuture::whenComplete

### ISSUE THREAD

**Issue by cretz** [typetools/checker-framework#1882]
I have the following code:

```java
package pkg;

import org.checkerframework.checker.nullness.qual.Nullable;
import java.util.concurrent.CompletableFuture;

public class Foo {
  public static CompletableFuture<@Nullable String> test(String param) {
    if ("foo".equals(param)) return CompletableFuture.completedFuture(null);
    return CompletableFuture.completedFuture("test").
        whenComplete((__, ___) -> System.out.println("Side effect"));
  }
}
```

It fails with:

    /redacted/src/main/java/pkg/Foo.java:11: error: [return.type.incompatible] incompatible types in return.
            whenComplete((__, ___) -> System.out.println("Side effect"));
                        ^
      found   : @Initialized @NonNull CompletableFuture<@Initialized @NonNull String>
      required: @Initialized @NonNull CompletableFuture<@Initialized @Nullable String>

It works fine without `whenComplete`. `whenComplete` should not change any typing.

EDIT: this seems to also occur with `thenCompose` and the like

**Comment by uemarica** [comment]
I don't think the return type of the test(String abc) method needs an Annotation of @Nullable(CompletableFuture<@Nullable String>),without it if you apply the nullness checker then only the if statement gives an error as: 
Error: [return.type.incompatible] incompatible types in return.
   found   : @Initialized @NonNull CompletableFuture<@Initialized @Nullable String>
   required: @Initialized @NonNull CompletableFuture<@Initialized @NonNull String>
which can be resolved by adding annotation @Nullable to the method parameter of completedFuture() method of the CompletableFuture class.

**Comment by uemarica** [comment]
I don't think the return type of the test(String abc) method needs an Annotation of @nullable(CompletableFuture<@nullable String>),without it if you apply the nullness checker then only the if statement gives an error as:
Error: [return.type.incompatible] incompatible types in return.
found : @initialized @nonnull CompletableFuture<@initialized @nullable String>
required: @initialized @nonnull CompletableFuture<@initialized @nonnull String>
which can be resolved by adding annotation @nullable to the method parameter of completedFuture() method of the CompletableFuture class.
I did it by creating a stub file of the CompletableFuture class.

**Comment by cretz** [comment]
Right...it's a problem with the annotated JDK. I am using http://repo1.maven.org/maven2/org/checkerframework/jdk8/2.4.0/ and I figured here was where you reported issues with the annotated JDK. I couldn't really tell from the docs whether I am expected to annotate JDK libs myself or utilize that package.

**Comment by mernst** [comment]
I'm sorry that was not clear.  (And thanks for the bug report.)

You should use the Maven Central annotated JDK when compiling.
If you discover a missing JDK annotation, then please submit a pull request that adds the missing annotations to some file under `checker/jdk`.  After the pull request is merged, we will update the Maven Central annotated JDK.

Could you make a suggestion of how the documentation could be improved to clarify this?
https://checkerframework.org/manual/#reporting-bugs does say:
"Please do not file an issue to report that some library class is not annotated. Instead, please write the annotations (see Chapter 29) and then submit a pull request."

**Comment by cretz** [comment]
Ah...no, that's plenty clear, I just missed it. But from a naive developer's point of view, most are not going to submit PRs just to get null checking right on common JDK classes like CompletableFuture. I understand the task of annotated the JDK is hard work, but from the outside the value of the null checker is minimal if developers have to do all of this. Maybe a caveat somewhere in the general part of the manual making it clear the JDK classes are not fully annotated and you will/might experience false positives that you have to ignore or write stubs to work around.

> Please do not file an issue to report that some library class is not annotated

At first, I thought it was just annotated wrong. How do I even know a class is not annotated? I have to disassemble the jdk8 JAR?

**Comment by mernst** [comment]
Thanks for the comments.  I have written a new section of the manual, which I will paste below.  Would this have been useful to answer your questions, and do you have any other suggestions for it?  Thanks again!

### 33.2.1  How to report problems with annotated libraries

Sometimes, a checker will report a warning because a library is not annotated or because the library contains incorrect annotations.

If the library is annotated but has some incorrect or missing annotations, then please open a pull request that adds the annotations. You can find the annotated JDK source in the Checker Framework GitHub repository, in directory checker/jdk. You can find the annotated source for other libraries in https://github.com/typetools/.

If the library is not annotated at all, you can fork it, add annotations, and propose that your annotated version be moved into https://github.com/typetools/.

For either approach, please see Chapter 29 for tips about writing library annotations.

Thanks for your contribution to the annotated libraries!

(Please don’t open an issue tracker issue just to report that some library class is not annotated. Such issues will be closed, to avoid cluttering the issue tracker.)

An alternative to adding annotations to a library is to write stub files, as described in Section 29.3. This is a quick and expedient way to add or chnage the annotations of an arbirary library. Creating an annotated library is a bit more work, but it is less error-prone, easier to use, runs slightly faster, and it can help other users.

**Comment by cretz** [comment]
Well, since you're asking, still a bit obtuse...sorry... :-)

> You can find the annotated JDK source in the Checker Framework GitHub repository, in directory checker/jdk.

Why not link https://github.com/typetools/checker-framework/tree/master/checker/jdk directly fi you are already linking the org? Concerned about linking the master branch? Also, the annotated source is under the sub directories (though kinda obvious). Also, lots of words there. I'd just keep it simple and specific to the JDK problem:

### False Positives on JDK Libraries

We try to maintain an up-to-date version of the JDK source with Checker Framework annotations applied for most classes at https://github.com/typetools/checker-framework/tree/master/checker/jdk. However, sometimes we don't have one that you might use causing false positives. Please do not open GitHub issues to have these handled. Instead, we appreciate pull requests to help us build the annotated source. Alternatively or in the interim, stub files can be provided as described in \<a>Section 29.3\</a>.

(end)

Not that big of a deal of course, just tossing my two cents in. I'll close this issue however since it was opened erroneously.

**Comment by mernst** [comment]
I agree about links to directories.  The references to directories were already links in the manual, but weren't retained when pasting into GitHub issues.  I added "subdirectories of" to prevent confusion.

Your version is short and to the point, which is good.  I'll adopt its wording about stub files, thanks!
The manual needs to cover cover non-JDK libraries, JDK annotations that appear as stub files, and it needs to cover false negatives (not to mention defining "false positive" and linking to the tips for annotation.  So I've shortened the text a bit, but retained important information.

Thanks again for the suggestions.
